### PR TITLE
test: destroy

### PR DESCRIPTION
**Pull request by rflechtner**
## re/ https://github.com/KILTprotocol/ticket/issues/3671

Tests the three destruction extrinsics. All tests passing rn.

To be rebased after merging #791.

## Checklist:

- [x] I have verified that the code works
  - [x] No panics! (checked arithmetic ops, no indexing `array[3]` use `get(3)`, ...)
- [x] I have verified that the code is easy to understand
  - [x] If not, I have left a well-balanced amount of inline comments
- [x] I have [left the code in a better state](https://deviq.com/principles/boy-scout-rule)
- [x] I have documented the changes (where applicable)
    * Either PR or Ticket to update [the Docs](https://github.com/KILTprotocol/docs)
    * Link the PR/Ticket here

### CODE REVIEW

File: pallets/pallet-bonded-coins/src/tests/transactions/finish_destroy.rs
Comment (Ad96el) [inline pallets/pallet-bonded-coins/src/tests/transactions/finish_destroy.rs:1]: I see an edge case regarding the deposits. If the deposit less the ED, it might be that the user got dusted, resulting into a failure to refund back his deposit. The pool life cycle would then be stuck. 

It can be avoided in the runtime, when the base deposit is equal or higher to the ED. Do we care about this test case?

File: pallets/pallet-bonded-coins/src/tests/transactions/finish_destroy.rs [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+
+	ExtBuilder::default()
+		.with_pools(vec![(pool_id.clone(), pool_details.clone())])
+		.with_native_balances(vec![(ACCOUNT_00, ONE_HUNDRED_KILT)])
+		.with_collaterals(vec![DEFAULT_COLLATERAL_CURRENCY_ID])
+		.with_bonded_balance(vec![(DEFAULT_BONDED_CURRENCY_ID, ACCOUNT_00, 100_000)])
+		.build()
+		.execute_with(|| {
+			// Fails because asset is not in destroying state
+			assert!(BondingPallet::finish_destroy(origin.clone(), pool_id.clone(), 1).is_err());
```
Comment (Ad96el) [inline pallets/pallet-bonded-coins/src/tests/transactions/finish_destroy.rs:212]: `assert_eq(BondingPallet::.finish_destroy(origin.clone(), pool_id.clone(), 1), Err(_));` is not possible?

File: pallets/pallet-bonded-coins/src/tests/transactions/start_destroy.rs [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+		.with_collaterals(vec![DEFAULT_COLLATERAL_CURRENCY_ID])
+		.build()
+		.execute_with(|| {
+			let origin = RawOrigin::Signed(ACCOUNT_00).into();
+
+			assert_ok!(BondingPallet::start_destroy(origin, pool_id.clone(), 1));
+
+			// Verify that the pool state has been updated to destroying
+			let updated_pool = Pools::<Test>::get(&pool_id).unwrap();
+			assert!(matches!(updated_pool.state, PoolStatus::Destroying));
```
Comment (Ad96el) [inline pallets/pallet-bonded-coins/src/tests/transactions/start_destroy.rs:35]: assert_eq! would look simpler. Same for the others further below

**Review by Ad96el — COMMENTED** [review]
LGTM. One style suggestion.

File: pallets/pallet-bonded-coins/src/tests/transactions/finish_destroy.rs (hunk 1 above)
Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/tests/transactions/finish_destroy.rs:212]: I think we have to use assert_eq. We introduced a neew lint, to avoid blank assert on results.

Comment (rflechtner) [reply pallets/pallet-bonded-coins/src/tests/transactions/finish_destroy.rs:212]: I think assert_eq only works for tests for equality, which don't allow matching patterns afaik. Same for assert_err!, which is why I'm not using that here. We can use a `matches!` expression, but the result is exactly the same as using the `is_err()`.

File: pallets/pallet-bonded-coins/src/tests/transactions/finish_destroy.rs
Comment (rflechtner) [reply pallets/pallet-bonded-coins/src/tests/transactions/finish_destroy.rs:1]: The deposit is just reserved, never actually removed from the account, is it not?

File: pallets/pallet-bonded-coins/src/tests/transactions/finish_destroy.rs (hunk 1 above)
Comment (rflechtner) [reply pallets/pallet-bonded-coins/src/tests/transactions/finish_destroy.rs:207]: I don't know, which one do you prefer?

Comment (Ad96el) [reply pallets/pallet-bonded-coins/src/tests/transactions/finish_destroy.rs:207]: Then leave it like that :+1:

File: pallets/pallet-bonded-coins/src/tests/transactions/start_destroy.rs (hunk 2 above)
Comment (rflechtner) [reply pallets/pallet-bonded-coins/src/tests/transactions/start_destroy.rs]: 1db6b8d7c70708533a100d0e28721449816fbe1a
